domain/miningmanager/miningmanager.go

```diff
@@ -19,16 +22,69 @@ type MiningManager interface {
 }
 
 type miningManager struct {
+	consensusReference   consensusreference.ConsensusReference
 	mempool              miningmanagermodel.Mempool
 	blockTemplateBuilder miningmanagermodel.BlockTemplateBuilder
+	cachedBlockTemplate  *externalapi.DomainBlockTemplate
+	cachingTime          time.Time
+	cacheLock            *sync.Mutex
 }
 
-// GetBlockTemplate creates a block template for a miner to consume
+// GetBlockTemplate obtains a block template for a miner to consume
 func (mm *miningManager) GetBlockTemplate(coinbaseData *externalapi.DomainCoinbaseData) (*externalapi.DomainBlock, error) {
-	return mm.blockTemplateBuilder.GetBlockTemplate(coinbaseData)
+	immutableCachedTemplate := mm.getImmutableCachedTemplate()
+	// We first try and use a cached template
+	if immutableCachedTemplate != nil {
+		virtualInfo, err := mm.consensusReference.Consensus().GetVirtualInfo()
+		if err != nil {
+			return nil, err
+		}
+		if externalapi.HashesEqual(virtualInfo.ParentHashes, immutableCachedTemplate.Block.Header.DirectParents()) {
+			if immutableCachedTemplate.CoinbaseData.Equal(coinbaseData) {
+				// Both, virtual parents and coinbase data are equal, simply return the cached block
+				return immutableCachedTemplate.Block, nil
+			}
+
+			// Virtual parents are equal, but coinbase data is new -- make the minimum changes required
+			// Note we first clone the block template since it is modified by the call
+			modifiedBlockTemplate, err := mm.blockTemplateBuilder.ModifyBlockTemplate(coinbaseData, immutableCachedTemplate.Clone())
+			if err != nil {
+				return nil, err
+			}
+
+			// No point in updating cache since we have no reason to believe this coinbase will be used more
+			// than the previous one, and we want to maintain the original template caching time
+			return modifiedBlockTemplate.Block, nil
+		}
+	}
+	// No relevant cache, build a template
+	blockTemplate, err := mm.blockTemplateBuilder.GetBlockTemplate(coinbaseData)
+	if err != nil {
+		return nil, err
+	}
+	// Cache the built template
+	mm.setImmutableCachedTemplate(blockTemplate)
+	return blockTemplate.Block, err
+}
+
+func (mm *miningManager) getImmutableCachedTemplate() *externalapi.DomainBlockTemplate {
+	mm.cacheLock.Lock()
+	defer mm.cacheLock.Unlock()
+	if time.Now().Sub(mm.cachingTime) > time.Second {
```

Comment:
Please add explanation. Is the only apparent reason to refresh the mempool? If yes, write it.